File: plugins/grok/scripts/lib/git.mjs
import path from "node:path";
import { runCommand } from "./process.mjs";
import { resolveWorkspaceRoot } from "./workspace.mjs";

const DEFAULT_MAX_DIFF_CHARS = 40_000;
const SMART_DIFF_THRESHOLD = 24_000;
const MAX_FILES_LISTED = 80;
const MAX_SAMPLED_FILES = 12;
const MAX_PER_FILE_PATCH = 4_000;

function git(cwd, args) {
  return runCommand("git", args, { cwd, timeoutMs: 30_000 });
}

export function ensureGitRepository(cwd) {
  const result = git(cwd, ["rev-parse", "--is-inside-work-tree"]);
  if (result.status !== 0 || result.stdout.trim() !== "true") {
    throw new Error("Not inside a git repository. Run this from a repo checkout.");
  }
  return resolveWorkspaceRoot(cwd);
}

/**
 * @param {string} cwd
 * @param {{ base?: string | null, scope?: string | null }} [options]
 */
export function resolveReviewTarget(cwd, options = {}) {
  const repoRoot = ensureGitRepository(cwd);
  const scope = options.scope ?? "auto";
  const base = options.base ?? null;

  if (base || scope === "branch") {
    const baseRef = base || "main";
    const shortstat = git(repoRoot, ["diff", "--shortstat", `${baseRef}...HEAD`]);
    const status = git(repoRoot, ["status", "--short", "--untracked-files=all"]);
    return {
      kind: "branch",
      base: baseRef,
      label: `branch vs ${baseRef}`,
      repoRoot,
      empty:
        !shortstat.stdout.trim() &&
        !status.stdout.trim() &&
        shortstat.status === 0
    };
  }

  const status = git(repoRoot, ["status", "--short", "--untracked-files=all"]);
  const cached = git(repoRoot, ["diff", "--shortstat", "--cached"]);
  const unstaged = git(repoRoot, ["diff", "--shortstat"]);
  const empty = !status.stdout.trim() && !cached.stdout.trim() && !unstaged.stdout.trim();

  return {
    kind: "working-tree",
    base: null,
    label: "working tree",
    repoRoot,
    empty
  };
}

/**
 * Collect a text summary of the review target for prompt injection.
 * @param {string} cwd
 * @param {ReturnType<typeof resolveReviewTarget>} target
 * @param {{ maxDiffChars?: number }} [options]
 */
export function collectReviewContext(cwd, target, options = {}) {
  const repoRoot = target.repoRoot || resolveWorkspaceRoot(cwd);
  const branchResult = git(repoRoot, ["rev-parse", "--abbrev-ref", "HEAD"]);
  const branch = branchResult.stdout.trim() || "HEAD";

  let diff = "";
  let status = "";
  let summary = "";
  let nameStatus = "";

  if (target.kind === "branch" && target.base) {
    const log = git(repoRoot, ["log", "--oneline", `${target.base}...HEAD`, "-20"]);
    const shortstat = git(repoRoot, ["diff", "--shortstat", `${target.base}...HEAD`]);
    const fullDiff = git(repoRoot, ["diff", `${target.base}...HEAD`]);
    const names = git(repoRoot, ["diff", "--name-status", `${target.base}...HEAD`]);
    status = log.stdout.trim();
    summary = shortstat.stdout.trim() || "No branch diff.";
    diff = fullDiff.stdout;
    nameStatus = names.stdout;
  } else {
    const st = git(repoRoot, ["status", "--short", "--untracked-files=all"]);
    const cached = git(repoRoot, ["diff", "--cached"]);
    const unstaged = git(repoRoot, ["diff"]);
    const namesC = git(repoRoot, ["diff", "--name-status", "--cached"]);
    const namesU = git(repoRoot, ["diff", "--name-status"]);
    status = st.stdout.trim();
    summary = [
      git(repoRoot, ["diff", "--shortstat", "--cached"]).stdout.trim(),
      git(repoRoot, ["diff", "--shortstat"]).stdout.trim()
    ]
      .filter(Boolean)
      .join(" | ") || (status ? "Untracked or status-only changes present." : "No changes.");
    diff = [cached.stdout, unstaged.stdout].filter(Boolean).join("\n");
    nameStatus = [namesC.stdout, namesU.stdout].filter(Boolean).join("\n");
  }

  const maxDiffChars = Number(options.maxDiffChars) || DEFAULT_MAX_DIFF_CHARS;
  diff = buildSmartDiff(diff, nameStatus, maxDiffChars);

  return {
    repoRoot,
    branch,
    target,
    status,
    summary,
    diff,
    relativeHint: path.relative(process.cwd(), repoRoot) || "."
  };
}

/**
 * @param {string} rawDiff
 * @param {string} nameStatus
 * @param {number} maxDiffChars
 */
export function buildSmartDiff(rawDiff, nameStatus, maxDiffChars = DEFAULT_MAX_DIFF_CHARS) {
  if (!rawDiff) {
    return "";
  }
  if (rawDiff.length <= Math.min(SMART_DIFF_THRESHOLD, maxDiffChars)) {
    return hardCap(rawDiff, maxDiffChars);
  }

  const files = String(nameStatus || "")
    .split(/\r?\n/)
    .map((l) => l.trim())
    .filter(Boolean);

  let out = `[smart-diff] Full diff is large (${rawDiff.length} chars). Showing name-status + sampled patches.\n\n`;
  out += `### Changed files (${files.length})\n`;
  for (const line of files.slice(0, MAX_FILES_LISTED)) {
    out += `${line}\n`;
  }
  if (files.length > MAX_FILES_LISTED) {
    out += `... ${files.length - MAX_FILES_LISTED} more files omitted\n`;
  }
  out += "\n";

  const patches = splitGitPatches(rawDiff);
  out += `### Sampled patches (first ${Math.min(MAX_SAMPLED_FILES, patches.length)} of ${patches.length})\n`;
  for (const patch of patches.slice(0, MAX_SAMPLED_FILES)) {
    let p = patch;
    if (p.length > MAX_PER_FILE_PATCH) {
      p = `${p.slice(0, MAX_PER_FILE_PATCH)}\n...[file patch truncated]...\n`;
    }
    out += `${p}\n`;
    if (out.length >= maxDiffChars) {
      break;
    }
  }
  return hardCap(out, maxDiffChars);
}

function splitGitPatches(diff) {
  if (!diff.includes("diff --git ")) {
    return [diff];
  }
  const parts = diff.split(/(?=^diff --git )/m).filter(Boolean);
  return parts.length ? parts : [diff];
}

function hardCap(s, max) {
  if (s.length <= max) {
    return s;
  }
  return `${s.slice(0, max)}\n\n...[diff truncated at ${max} chars]...`;
}

File: crates/grok-companion/src/git.rs
use std::path::{Path, PathBuf};

use crate::error::{CompanionError, Result};
use crate::process::run_command;
use crate::workspace::resolve_workspace_root;

const MAX_DIFF_CHARS: usize = 120_000;

#[derive(Debug, Clone)]
pub struct ReviewTarget {
    pub kind: String, // "working-tree" | "branch"
    pub base: Option<String>,
    pub label: String,
    pub repo_root: PathBuf,
    pub empty: bool,
}

#[derive(Debug, Clone)]
pub struct ReviewContext {
    pub repo_root: PathBuf,
    pub branch: String,
    pub target: ReviewTarget,
    pub status: String,
    pub summary: String,
    pub diff: String,
}

fn git(cwd: &Path, args: &[&str]) -> Result<(i32, String, String)> {
    let result = run_command("git", args, Some(cwd), None)?;
    Ok((result.status, result.stdout, result.stderr))
}

pub fn ensure_git_repository(cwd: &Path) -> Result<PathBuf> {
    let (status, stdout, stderr) = git(cwd, &["rev-parse", "--is-inside-work-tree"])?;
    if status != 0 || stdout.trim() != "true" {
        return Err(CompanionError::msg(format!(
            "Not inside a git repository. {stderr}"
        )));
    }
    Ok(resolve_workspace_root(cwd))
}

pub fn resolve_review_target(
    cwd: &Path,
    base: Option<&str>,
    scope: Option<&str>,
) -> Result<ReviewTarget> {
    let repo_root = ensure_git_repository(cwd)?;
    let scope = scope.unwrap_or("auto");

    if base.is_some() || scope == "branch" {
        let base_ref = base.unwrap_or("main");
        let (st_stat, shortstat, _) = git(
            &repo_root,
            &["diff", "--shortstat", &format!("{base_ref}...HEAD")],
        )?;
        let (_, status, _) = git(
            &repo_root,
            &["status", "--short", "--untracked-files=all"],
        )?;
        let empty = shortstat.trim().is_empty() && status.trim().is_empty() && st_stat == 0;
        return Ok(ReviewTarget {
            kind: "branch".into(),
            base: Some(base_ref.into()),
            label: format!("branch vs {base_ref}"),
            repo_root,
            empty,
        });
    }

    let (_, status, _) = git(
        &repo_root,
        &["status", "--short", "--untracked-files=all"],
    )?;
    let (_, cached, _) = git(&repo_root, &["diff", "--shortstat", "--cached"])?;
    let (_, unstaged, _) = git(&repo_root, &["diff", "--shortstat"])?;
    let empty =
        status.trim().is_empty() && cached.trim().is_empty() && unstaged.trim().is_empty();

    Ok(ReviewTarget {
        kind: "working-tree".into(),
        base: None,
        label: "working tree".into(),
        repo_root,
        empty,
    })
}

pub fn collect_review_context(cwd: &Path, target: &ReviewTarget) -> Result<ReviewContext> {
    let repo_root = if target.repo_root.as_os_str().is_empty() {
        resolve_workspace_root(cwd)
    } else {
        target.repo_root.clone()
    };

    let (_, branch_out, _) = git(&repo_root, &["rev-parse", "--abbrev-ref", "HEAD"])?;
    let branch = {
        let b = branch_out.trim();
        if b.is_empty() {
            "HEAD".into()
        } else {
            b.to_string()
        }
    };

    let (status, summary, mut diff) = if target.kind == "branch" {
        let base = target.base.as_deref().unwrap_or("main");
        let (_, log, _) = git(
            &repo_root,
            &["log", "--oneline", &format!("{base}...HEAD"), "-20"],
        )?;
        let (_, shortstat, _) =
            git(&repo_root, &["diff", "--shortstat", &format!("{base}...HEAD")])?;
        let (_, full_diff, _) = git(&repo_root, &["diff", &format!("{base}...HEAD")])?;
        let summary = if shortstat.trim().is_empty() {
            "No branch diff.".into()
        } else {
            shortstat.trim().to_string()
        };
        (log.trim().to_string(), summary, full_diff)
    } else {
        let (_, st, _) = git(
            &repo_root,
            &["status", "--short", "--untracked-files=all"],
        )?;
        let (_, cached, _) = git(&repo_root, &["diff", "--cached"])?;
        let (_, unstaged, _) = git(&repo_root, &["diff"])?;
        let (_, cached_ss, _) = git(&repo_root, &["diff", "--shortstat", "--cached"])?;
        let (_, unstaged_ss, _) = git(&repo_root, &["diff", "--shortstat"])?;
        let parts: Vec<&str> = [cached_ss.trim(), unstaged_ss.trim()]
            .into_iter()
            .filter(|s| !s.is_empty())
            .collect();
        let summary = if parts.is_empty() {
            if st.trim().is_empty() {
                "No changes.".into()
            } else {
                "Untracked or status-only changes present.".into()
            }
        } else {
            parts.join(" | ")
        };
        let mut d = String::new();
        if !cached.is_empty() {
            d.push_str(&cached);
        }
        if !unstaged.is_empty() {
            if !d.is_empty() {
                d.push('\n');
            }
            d.push_str(&unstaged);
        }
        (st.trim().to_string(), summary, d)
    };

    if diff.len() > MAX_DIFF_CHARS {
        diff = format!(
            "{}\n\n...[diff truncated at {MAX_DIFF_CHARS} chars]...",
            &diff[..MAX_DIFF_CHARS]
        );
    }

    Ok(ReviewContext {
        repo_root,
        branch,
        target: target.clone(),
        status,
        summary,
        diff,
    })
}
